rts/motoko-rts/src/stabilization.rs

```diff
@@ -0,0 +1,74 @@
+//! Graph-copy-based stabilzation on upgrades, serializing the entire stable object graph into
+//! stable memory by using a defined long-term stable storage format.
+//!
+//! This is to support potentially radical changes of the persistent main memory layout, e.g.
+//! introducing a new GC or rearranging persistent metadata. This also relies on precise value
+//! tagging to allow more advanced changes that require value metadata, e.g. specializing arrays for
+//! small element types or even downgrading to 32-bit heap layouts (provided that the amount of live
+//! data fits into a 32-bit memory).
+//!
+//! A memory compatibility check similar to enhanced orthogonal persistence has to be performed.
+//! For this purpose, the type table of the serialized object graph is also stored in stable memory
+//! and on upgrade, compared to the new program version.
+//!
+//! A versioned stable storage format even permits future evolutions of the graph copy algorithm.
+//!  
+//! See `GraphCopyStabilization.md` for the stable format specification and the employed algorithm.
+
+pub mod deserialization;
+pub mod graph_copy;
+pub mod layout;
+pub mod serialization;
+
+#[cfg(feature = "ic")]
+pub mod ic;
+
+use core::cmp::min;
+
+use crate::{
+    constants::WASM_PAGE_SIZE,
+    rts_trap_with,
+    stable_mem::{self, ic0_stable64_write, PAGE_SIZE},
+};
+
+use self::layout::StableValue;
+
+extern "C" {
+    pub fn moc_stabilization_instruction_limit() -> u64;
+    fn ic0_performance_counter(number: u32) -> u64;
+}
+
+// Dummy value used for non-stable objects that are potentially reachable from
+// stable variable because of structural subtyping or `Any`-subtyping.
+// Must be a non-skewed value such that the GC also ignores this value.
+const DUMMY_VALUE: StableValue = StableValue::from_raw(0);
+
+fn clear_stable_memory(start: u64, length: u64) {
+    const CHUNK_SIZE: usize = WASM_PAGE_SIZE.as_usize();
+    let empty_chunk = [0u8; CHUNK_SIZE];
+    let mut position = start;
+    let end = start + length;
+    while position < end {
+        let size = min(end - position, CHUNK_SIZE as u64);
+        unsafe {
+            ic0_stable64_write(position, &empty_chunk as *const u8 as u64, size);
+        }
+        position += size;
+    }
+}
+
+fn grant_stable_space(byte_size: u64) {
+    debug_assert!(byte_size < u64::MAX - PAGE_SIZE - 1);
+    let required_pages = (byte_size + PAGE_SIZE - 1) / PAGE_SIZE;
+    let available_pages = stable_mem::size();
```

Comment:
I forget, are these the operations for growing physical or virtual stable memory. Will need to check.

Reply:
The fix might be to call ic0_stable64_grow and and ic_stable64_size instead.

Reply:
Maybe it isn't a bug.

I just ran this test, that allocates 3 stable memory pages and the count is preserved across upgrades.
```
//MOC-FLAG --stabilization-instruction-limit=10000

import Prim "mo:prim";

actor {

    if (Prim.stableMemorySize() == 0) {
      Prim.debugPrint("growing stable memory");
      ignore Prim.stableMemoryGrow(3);
    };
    assert Prim.stableMemorySize() == 3;
    Prim.debugPrint(debug_show {stableMemorySize = Prim.stableMemorySize()});
    stable var largeArray = Prim.Array_tabulate<Nat>(100_000, func(index) { index });

    public func check() : async () {
        Prim.debugPrint("Array of length " # debug_show (largeArray.size()));
        var index = 0;
        while (index < largeArray.size()) {
            assert (largeArray[index] == index);
            index += 1;
        };
    };

    system func preupgrade() {
        Prim.debugPrint("PRE-UPGRADE HOOK!");
    };

    system func postupgrade() {
        Prim.debugPrint("POST-UPGRADE HOOK!");
    };
};

//CALL ingress check "DIDL\x00\x00"
//CALL ingress __motoko_stabilize_before_upgrade "DIDL\x00\x00"
//CALL upgrade ""
//CALL ingress __motoko_destabilize_after_upgrade "DIDL\x00\x00"
//CALL ingress check "DIDL\x00\x00"
//CALL ingress __motoko_stabilize_before_upgrade "DIDL\x00\x00"
//CALL upgrade ""
//CALL ingress __motoko_destabilize_after_upgrade "DIDL\x00\x00"
//CALL ingress check "DIDL\x00\x00"

//SKIP run
//SKIP run-ir
//SKIP run-low
```
output:
```
[nix-shell:~/clean/motoko/test/run-drun]$ ../run.sh -d stabilize-large-array-esm.mo
WARNING: Could not run ic-ref-run, will skip running some tests
stabilize-large-array-esm: [tc] [comp] [comp-ref] [valid] [valid-ref] [drun-run]
--- stabilize-large-array-esm.drun-run (expected)
+++ stabilize-large-array-esm.drun-run (actual)
@@ -0,0 +1,22 @@
+ingress Completed: Reply: 0x4449444c016c01b3c4b1f204680100010a00000000000000000101
+debug.print: growing stable memory
+debug.print: {stableMemorySize = 3}
+ingress Completed: Reply: 0x4449444c0000
+debug.print: Array of length 100_000
+ingress Completed: Reply: 0x4449444c0000
+debug.print: PRE-UPGRADE HOOK!
+ingress Completed: Reply: 0x4449444c0000
+ingress Completed: Reply: 0x4449444c0000
+debug.print: {stableMemorySize = 3}
+debug.print: POST-UPGRADE HOOK!
+ingress Completed: Reply: 0x4449444c0000
+debug.print: Array of length 100_000
+ingress Completed: Reply: 0x4449444c0000
+debug.print: PRE-UPGRADE HOOK!
+ingress Completed: Reply: 0x4449444c0000
+ingress Completed: Reply: 0x4449444c0000
+debug.print: {stableMemorySize = 3}
+debug.print: POST-UPGRADE HOOK!
+ingress Completed: Reply: 0x4449444c0000
+debug.print: Array of length 100_000
+ingress Completed: Reply: 0x4449444c0000
```

Maybe its fine because you write the ESM page count to metadata before doing the actual graph copy.

Reply:
(But I haven't verified that, it's my guess at why this still works)

Reply:
Ok, turns out this is a real bug because stablemem_grow, unlike ic0_stable64_grow, caps growth to --max-stable-pages, and will prevent stabilization of larger heap data.

See #4540

Reply:
Thanks a lot for having noticed this bug and all the analysis and the repro.
It should now be fixed, using physical memory size and grow.